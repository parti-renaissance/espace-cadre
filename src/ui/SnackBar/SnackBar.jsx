import { forwardRef, useCallback, useState } from 'react'
import PropTypes from 'prop-types'
import { SnackbarContent } from 'notistack'
import { Card, CardActions as MuiCardActions, Collapse, Grid, IconButton, Paper, Typography } from '@mui/material'
import { styled } from '@mui/system'
import CloseIcon from '@mui/icons-material/Close'
import ExpandMoreIcon from '@mui/icons-material/ExpandMore'
import { shouldForwardProps } from '~/components/shared/shouldForwardProps'
import { useCustomSnackbar } from '~/components/shared/notification/hooks'

const Wrapper = styled(SnackbarContent)`
  min-width: 300px;
  max-width: 400px;
  display: flex;
`
const CardActions = styled(MuiCardActions)(
  ({ theme, variant }) => `
  padding: ${theme.spacing(1, 1, 1, 2)};
  justify-content: space-between;
  background: ${theme.palette.notification[variant].background};
  color: ${theme.palette.whiteCorner};
`
)
const Title = styled(props => <Typography variant="body2" {...props} />, shouldForwardProps)`
  padding: ${({ theme }) => theme.spacing(0, 3, 0, 1)};
`
const ExpandButton = styled(
  IconButton,
  shouldForwardProps
)(
  ({ theme, expanded }) => `
transform: rotate(${expanded ? 180 : 0}deg);
transition: ${theme.transitions.create('transform', { duration: theme.transitions.duration.shortest })};
`
)

const ColoredExpandIcon = styled(ExpandMoreIcon)`
  color: ${({ theme }) => theme.palette.whiteCorner};
`
const ColoredCloseIcon = styled(CloseIcon)`
  color: ${({ theme }) => theme.palette.whiteCorner};
`
const Content = styled(
  Paper,
  shouldForwardProps
)(
  ({ theme, bold }) => `
  padding: ${theme.spacing(2, 2, 2, 3)};
  background: ${theme.palette.whiteCorner};
  color: ${theme.palette.blackCorner};
  border: 1px solid ${theme.palette.gray300};
  font-weight: ${bold === true ? 600 : 'initial'};
  font-size: 14px;
`
)

const UISnackBar = forwardRef(({ id, message, variant, dismissResolve, children }, ref) => {
  const [expanded, setExpanded] = useState(false)
  const { closeSnackbar } = useCustomSnackbar()

  const handleExpandToggle = useCallback(() => {
    setExpanded(currentExpanded => !currentExpanded)
  }, [])

  const handleDismiss = useCallback(() => {
    closeSnackbar(id)
    dismissResolve && dismissResolve()
  }, [id, closeSnackbar, dismissResolve])

  return (
    <Wrapper ref={ref}>
      <Card>
        <CardActions variant={variant}>
          <Grid item>
            <Title>{message}</Title>
          </Grid>
          <Grid item>
            {children && (
              <ExpandButton expanded={expanded} onClick={handleExpandToggle}>
                <ColoredExpandIcon fontSize="small" />
              </ExpandButton>
            )}
            <IconButton onClick={handleDismiss}>
              <ColoredCloseIcon fontSize="small" />
            </IconButton>
          </Grid>
        </CardActions>
        {children && (
          <Collapse in={expanded} timeout="auto">
            <Content bold>{children}</Content>
          </Collapse>
        )}
      </Card>
    </Wrapper>
  )
})

UISnackBar.displayName = 'UISnackBar'
UISnackBar.defaultProps = {
  content: null,
  dismissResolve: null,
  children: null,
}
UISnackBar.propTypes = {
  id: PropTypes.string.isRequired,
  message: PropTypes.string.isRequired,
  variant: PropTypes.string.isRequired,
  children: PropTypes.node,
  dismissResolve: PropTypes.func,
}

export default UISnackBar
