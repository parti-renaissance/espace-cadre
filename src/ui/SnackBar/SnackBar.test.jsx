import { fireEvent, render, screen } from '@testing-library/react'
import SnackBar from '~/ui/SnackBar/SnackBar'

vi.mock('react-redux', () => ({
  useDispatch: () => vi.fn(),
}))
vi.mock('notistack', () => ({
  SnackbarContent: ({ children }) => <div className="mock-snackbar-content">{children}</div>,
}))
vi.mock('@mui/system', () => ({
  styled: c => () => c,
}))
vi.mock('@mui/material', () => ({
  Card: ({ children }) => <div className="mock-card">{children}</div>,
  CardActions: ({ children, variant }) => <div className={`mock-card-actions ${variant}`}>{children}</div>,
  Grid: ({ children }) => <div className="mock-grid">{children}</div>,
  Typography: ({ children }) => <div className="mock-typography">{children}</div>,
  IconButton: ({ children, onClick }) => (
    <div className="mock-icon-button" data-testid="icon-button" onClick={onClick}>
      {children}
    </div>
  ),
  Collapse: ({ children, in: _in }) => <div className="mock-collapse">{_in && children}</div>,
  Paper: ({ children }) => <div className="mock-paper">{children}</div>,
}))
vi.mock('@mui/icons-material/Close', () => ({ default: () => <div className="mock-close-icon" /> }))
vi.mock('@mui/icons-material/ExpandMore', () => ({ default: () => <div className="mock-expand-more-icon" /> }))
const mockCloseSnackbar = vi.fn()
vi.mock('~/components/shared/notification/hooks', () => ({
  useCustomSnackbar: () => ({ closeSnackbar: mockCloseSnackbar }),
}))

describe('SnackBar', () => {
  it('displays a SnackBar', () => {
    const mockDismiss = vi.fn()
    const { container } = render(<SnackBar id="id" message="message" variant="variant" dismissResolve={mockDismiss} />)

    expect(container).toMatchSnapshot()
  })

  it('displays a SnackBar with content', () => {
    const mockDismiss = vi.fn()
    const { container } = render(
      <SnackBar id="id" message="message" variant="variant" dismissResolve={mockDismiss}>
        <div data-testid="content" />
      </SnackBar>
    )

    const [toggleContent] = screen.getAllByTestId('icon-button')
    fireEvent.click(toggleContent)
    expect(container).toMatchSnapshot()
  })

  it('calls handleDismiss callback', () => {
    const mockDismiss = vi.fn()
    render(
      <SnackBar id="id" message="message" variant="variant" dismissResolve={mockDismiss}>
        <div data-testid="content" />
      </SnackBar>
    )
    const [, dismissButton] = screen.getAllByTestId('icon-button')
    fireEvent.click(dismissButton)
    expect(mockDismiss).toHaveBeenCalled()
    expect(mockCloseSnackbar).toHaveBeenCalled()
  })

  it('displays/hides content on click', () => {
    const mockDismiss = vi.fn()
    render(
      <SnackBar id="id" message="message" variant="variant" dismissResolve={mockDismiss}>
        <div data-testid="content" />
      </SnackBar>
    )

    expect(screen.queryByTestId('content')).toBeFalsy()

    const [toggleContent] = screen.getAllByTestId('icon-button')
    fireEvent.click(toggleContent)
    screen.getByTestId('content')

    fireEvent.click(toggleContent)
    expect(screen.queryByTestId('content')).toBeFalsy()
  })
})
