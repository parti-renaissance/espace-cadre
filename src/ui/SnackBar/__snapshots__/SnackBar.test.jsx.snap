// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SnackBar > displays a SnackBar 1`] = `
<div>
  <div
    class="mock-snackbar-content"
  >
    <div
      class="mock-card"
    >
      <div
        class="mock-card-actions variant"
      >
        <div
          class="mock-grid"
        >
          <div
            class="mock-typography"
          >
            message
          </div>
        </div>
        <div
          class="mock-grid"
        >
          <div
            class="mock-icon-button"
            data-testid="icon-button"
          >
            <div
              class="mock-close-icon"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SnackBar > displays a SnackBar with content 1`] = `
<div>
  <div
    class="mock-snackbar-content"
  >
    <div
      class="mock-card"
    >
      <div
        class="mock-card-actions variant"
      >
        <div
          class="mock-grid"
        >
          <div
            class="mock-typography"
          >
            message
          </div>
        </div>
        <div
          class="mock-grid"
        >
          <div
            class="mock-icon-button"
            data-testid="icon-button"
          >
            <div
              class="mock-expand-more-icon"
            />
          </div>
          <div
            class="mock-icon-button"
            data-testid="icon-button"
          >
            <div
              class="mock-close-icon"
            />
          </div>
        </div>
      </div>
      <div
        class="mock-collapse"
      >
        <div
          class="mock-paper"
        >
          <div
            data-testid="content"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SnackBar displays a SnackBar 1`] = `
<div>
  <div
    class="mock-snackbar-content"
  >
    <div
      class="mock-card"
    >
      <div
        class="mock-card-actions variant"
      >
        <div
          class="mock-grid"
        >
          <div
            class="mock-typography"
          >
            message
          </div>
        </div>
        <div
          class="mock-grid"
        >
          <div
            class="mock-icon-button"
            data-testid="icon-button"
          >
            <div
              class="mock-close-icon"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SnackBar displays a SnackBar with content 1`] = `
<div>
  <div
    class="mock-snackbar-content"
  >
    <div
      class="mock-card"
    >
      <div
        class="mock-card-actions variant"
      >
        <div
          class="mock-grid"
        >
          <div
            class="mock-typography"
          >
            message
          </div>
        </div>
        <div
          class="mock-grid"
        >
          <div
            class="mock-icon-button"
            data-testid="icon-button"
          >
            <div
              class="mock-expand-more-icon"
            />
          </div>
          <div
            class="mock-icon-button"
            data-testid="icon-button"
          >
            <div
              class="mock-close-icon"
            />
          </div>
        </div>
      </div>
      <div
        class="mock-collapse"
      >
        <div
          class="mock-paper"
        >
          <div
            data-testid="content"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
